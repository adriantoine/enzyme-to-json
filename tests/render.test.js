/* eslint-env jest */

import React from 'react';
import Enzyme, {render} from 'enzyme';
import omitBy from 'lodash/omitBy';
import Adapter from 'enzyme-adapter-react-16';

import renderToJson from '../src/render';
import {
  BasicPure,
  BasicWithAList,
  ArrayRender,
  FragmentAsChild,
  FragmentAsRoot,
} from './fixtures/pure-function';
import {
  BasicClass,
  ClassWithNull,
  ClassWithNullChildren,
  ClassArrayRender,
} from './fixtures/class';

Enzyme.configure({adapter: new Adapter()});

it('doesnt break when called without arguments', () => {
  expect(renderToJson()).toBe(null);
  expect(renderToJson([null])).toBe(null);
  expect(renderToJson(['a'])).toBe(null);
});

it('converts basic pure render', () => {
  const rendered = render(
    <BasicPure className="pure">
      <strong>Hello!</strong>
    </BasicPure>,
  );

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('converts basic class render', () => {
  const rendered = render(
    <BasicClass className="class">
      <strong>Hello!</strong>
    </BasicClass>,
  );

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('renders the whole list', () => {
  const wrapper = render(<BasicWithAList />);
  expect(renderToJson(wrapper.find('ul'))).toMatchSnapshot();
});

it('handles a component which returns null', () => {
  const rendered = render(<ClassWithNull />);
  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('handles a component having null children', () => {
  const rendered = render(<ClassWithNullChildren />);
  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('renders multiple elements as a result of find', () => {
  const rendered = render(<BasicWithAList />);
  expect(renderToJson(rendered.find('li'))).toMatchSnapshot();
});

it('converts function components with render returning top level arrays', () => {
  const rendered = render(
    <ArrayRender>
      <strong>Hello!</strong>
    </ArrayRender>,
  );

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('converts class components with render returning top level arrays', () => {
  const rendered = render(
    <ClassArrayRender>
      <strong>Hello!</strong>
    </ClassArrayRender>,
  );

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('accepts a map option allowing to customize content', () => {
  const rendered = render(<strong>Hello!</strong>);

  expect(
    renderToJson(rendered, {
      map: json => ({...json, children: ['Goodbye!']}),
    }),
  ).toMatchSnapshot();
});

it('accepts a map option allowing to customize content of all nested components', () => {
  const rendered = render(
    <div randomlygeneratedkey={Date.now()} className="wrapper">
      <strong randomlygeneratedkey={Date.now()}>Hello!</strong>
      <strong className="strong2">Hello 2</strong>
    </div>,
  );

  expect(
    renderToJson(rendered, {
      map: json => ({
        ...json,
        props: omitBy(json.props, (val, key) => key === 'randomlygeneratedkey'),
      }),
    }),
  ).toMatchSnapshot();
});

it('can skip a component I dont want to see with the map option', () => {
  const rendered = render(
    <div>
      <ul>
        <li>Item 1</li>
        <li>Item 2</li>
      </ul>
      <strong>Hello 2</strong>
    </div>,
  );

  expect(
    renderToJson(rendered, {
      map: json => {
        if (json.type === 'ul') {
          return null;
        }
        return json;
      },
    }),
  ).toMatchSnapshot();
});

it('outputs the snapshot even with inline JSX conditions being falsy', () => {
  const rendered = render(
    <div>
      <span>I am there</span>
      {false && <span>Issue</span>}
    </div>,
  );

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('renders a component that has a child fragment', () => {
  const wrapper = render(<FragmentAsChild />);

  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('renders a component that has a fragment root', () => {
  const wrapper = render(<FragmentAsRoot />);

  expect(renderToJson(wrapper)).toMatchSnapshot();
});
