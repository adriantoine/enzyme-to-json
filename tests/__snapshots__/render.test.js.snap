// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts a map option allowing to customize content 1`] = `
<strong>
  Goodbye!
</strong>
`;

exports[`accepts a map option allowing to customize content of all nested components 1`] = `
<div
  class="wrapper"
>
  <strong>
    Hello!
  </strong>
  <strong
    class="strong2"
  >
    Hello 2
  </strong>
</div>
`;

exports[`can skip a component I dont want to see with the map option 1`] = `
<div>
  <strong>
    Hello 2
  </strong>
</div>
`;

exports[`converts basic class render 1`] = `
<div
  class="basic-class class"
>
  <div
    class="group"
    id="group-id"
  >
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      class="empty"
    />
  </div>
</div>
`;

exports[`converts basic pure render 1`] = `
<div
  class="basic-pure pure"
>
  <div
    class="group"
    id="group-id"
  >
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      class="empty"
    />
  </div>
</div>
`;

exports[`converts class components with render returning top level arrays 1`] = `
Array [
  <div
    class="test"
  >
    Test
  </div>,
  <div
    class="test2"
  >
    Test 2
  </div>,
  <div
    class="child"
  >
    <strong>
      Hello!
    </strong>
  </div>,
]
`;

exports[`converts function components with render returning top level arrays 1`] = `
Array [
  <div
    class="test"
  >
    Test
  </div>,
  <div
    class="test2"
  >
    Test 2
  </div>,
  <div
    class="child"
  >
    <strong>
      Hello!
    </strong>
  </div>,
]
`;

exports[`handles a component having null children 1`] = `<div />`;

exports[`handles a component which returns null 1`] = `null`;

exports[`outputs the snapshot even with inline JSX conditions being falsy 1`] = `
<div>
  <span>
    I am there
  </span>
</div>
`;

exports[`renders a component that has a child fragment 1`] = `
<div>
  <span />
  <div />
  <button />
</div>
`;

exports[`renders a component that has a fragment root 1`] = `
Array [
  <span />,
  <div />,
  <button />,
]
`;

exports[`renders multiple elements as a result of find 1`] = `
Array [
  <li>
    0
  </li>,
  <li>
    1
  </li>,
]
`;

exports[`renders the whole list 1`] = `
<ul>
  <li>
    0
  </li>
  <li>
    1
  </li>
</ul>
`;
